src/librustc_macros/src/query.rs

```diff
@@ -0,0 +1,302 @@
+            panic!("attribute `{}` not supported on queries", path.0)
+        }
+    }
+}
+
+impl Parse for Query {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        check_attributes(input.call(Attribute::parse_outer)?);
+
+        let query: Ident = input.parse()?;
+        if query != "query" {
+            panic!("expected `query`");
+        }
+        let name: Ident = input.parse()?;
+        let arg_content;
+        parenthesized!(arg_content in input);
+        let key = arg_content.parse()?;
+        arg_content.parse::<Token![:]>()?;
+        let arg = arg_content.parse()?;
+        if !arg_content.is_empty() {
+            panic!("expected only one query argument");
+        };
+        let result = input.parse()?;
+
+        let content;
+        braced!(content in input);
+        let attrs = content.parse()?;
+
+        Ok(Query {
+            attrs,
+            name,
+            key,
+            arg,
+            result,
+        })
+    }
+}
+
+struct List<T>(Vec<T>);
+
+impl<T: Parse> Parse for List<T> {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let mut list = Vec::new();
+        while !input.is_empty() {
+            list.push(input.parse()?);
+        }
+        Ok(List(list))
+    }
+}
+
+struct Group {
+    name: Ident,
+    queries: List<Query>,
+}
+
+impl Parse for Group {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let name: Ident = input.parse()?;
+        let content;
+        braced!(content in input);
+        Ok(Group {
+            name,
+            queries: content.parse()?,
+        })
+    }
+}
+
+fn camel_case(string: &str) -> String {
+    let mut pos = vec![0];
+    for (i, c) in string.chars().enumerate() {
+        if c == '_' {
+            pos.push(i + 1);
+        }
+    }
+    string.chars().enumerate().filter(|c| c.1 != '_').flat_map(|(i, c)| {
+        if pos.contains(&i) {
+            c.to_uppercase().collect::<Vec<char>>()
+        } else {
+            vec![c]
+        }
+    }).collect()
+}
+
+pub fn rustc_queries(input: TokenStream) -> TokenStream {
+    let groups = parse_macro_input!(input as List<Group>);
+
+    let mut query_stream = quote! {};
+    let mut query_description_stream = quote! {};
+    let mut dep_node_def_stream = quote! {};
+    let mut dep_node_force_stream = quote! {};
+
+    for group in groups.0 {
+        let mut group_stream = quote! {};
+        for query in &group.queries.0 {
+            let name = &query.name;
+            let dep_node_name = Ident::new(
+                &camel_case(&name.to_string()),
+                name.span());
+            let arg = &query.arg;
+            let key = &query.key.0;
+            let result_full = &query.result;
+            let result = match query.result {
+                ReturnType::Default => quote! { -> () },
+                _ => quote! { #result_full },
+            };
+
+            // Find out if we should cache the query on disk
```

Comment:
This is the only comment explaining what all the parsing code is doing. I think some more docs and comments are warranted given the complexity of it.

Reply:
I added some more of them.